src/vm.c

```diff
@@ -72,6 +73,53 @@ static void vm_init_cpu_id(vm_t *v)
     ioctl(v->vcpu_fd, KVM_SET_CPUID2, &kvm_cpuid);
 }
 
+#define FREQ_NS 1000000
+
+static void handler(int sig, siginfo_t *si, void *uc) {}
+
+static int vm_init_timer(vm_t *v)
+{
+    timer_t timerid;
+    struct sigevent sev;
+    struct itimerspec its;
+    struct sigaction sa;
+    sigset_t mask;
+
+    sa.sa_flags = SA_SIGINFO;
+    sa.sa_sigaction = handler;
+    sigemptyset(&sa.sa_mask);
+    if (sigaction(SIGALRM, &sa, NULL) == -1)
```

Comment:
How about using timerfd instead of `SIGALRM`?

Reply:
The advantage of `SIGALRM` is the signal to interrupt periodically, which can depend on a standard library / operating system directly without extra codes. If we want to use timerfd instead, we may need a thread to poll the file descriptor for whether it is readable, and trigger interrupts explicitly if the timer expires. Otherwise, VM will always run inside the loop of KVM_RUN ioctl and never return.

Reply:
Except we can have some good approaches to listen for the timerfd without other threads, I think `SIGALRM` is more simple since we don't actually need to do anything in the signal handler.

Reply:
Can you try the mixture of timerfd and epoll instead?

Reply:
Sure, maybe I can show how I think about the implementation using timerfd first. Then we can compare both methods.